dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/WorkflowItemRestRepository.java

```diff
@@ -0,0 +1,335 @@
+                    AbstractProcessingStep stepProcessing = (AbstractProcessingStep) stepClass.newInstance();
+                    stepProcessing.doPreProcessing(context, source);
+                } else {
+                    throw new Exception(
+                        "The submission step class specified by '" + stepConfig.getProcessingClassName() +
+                        "' does not extend the class org.dspace.submit.AbstractProcessingStep!" +
+                        " Therefore it cannot be used by the Configurable Submission as the <processing-class>!");
+                }
+
+            } catch (Exception e) {
+                log.error(e.getMessage(), e);
+            }
+        }
+        try {
+            submissionService.saveWorkflowItem(context, source);
+        } catch (SQLException | AuthorizeException e) {
+            throw new RuntimeException(e.getMessage(), e);
+        }
+        return wfi;
+    }
+
+    @Override
+    public Class<WorkflowItemRest> getDomainClass() {
+        return WorkflowItemRest.class;
+    }
+
+    @Override
+    public WorkflowItemResource wrapResource(WorkflowItemRest witem, String... rels) {
+        return new WorkflowItemResource(witem, utils, rels);
+    }
+
+    @Override
+    public WorkflowItemRest upload(HttpServletRequest request, String apiCategory, String model, Integer id,
+                                   MultipartFile file) throws Exception {
+
+        Context context = obtainContext();
+        WorkflowItemRest wsi = findOne(id);
+        XmlWorkflowItem source = wis.find(context, id);
+        List<ErrorRest> errors = new ArrayList<ErrorRest>();
+        SubmissionConfig submissionConfig =
+            submissionConfigReader.getSubmissionConfigByName(wsi.getSubmissionDefinition().getName());
+        for (int i = 0; i < submissionConfig.getNumberOfSteps(); i++) {
+            SubmissionStepConfig stepConfig = submissionConfig.getStep(i);
+
+            /*
+             * First, load the step processing class (using the current
+             * class loader)
+             */
+            ClassLoader loader = this.getClass().getClassLoader();
+            Class stepClass;
+            try {
+                stepClass = loader.loadClass(stepConfig.getProcessingClassName());
+
+                Object stepInstance = stepClass.newInstance();
+                if (UploadableStep.class.isAssignableFrom(stepClass)) {
+                    UploadableStep uploadableStep = (UploadableStep) stepInstance;
+                    uploadableStep.doPreProcessing(context, source);
+                    ErrorRest err =
+                        uploadableStep.upload(context, submissionService, stepConfig, source, file);
+                    uploadableStep.doPostProcessing(context, source);
+                    if (err != null) {
+                        errors.add(err);
+                    }
+                }
+
+            } catch (Exception e) {
+                log.error(e.getMessage(), e);
```

Comment:
Another area of catching an error in a for loop.  Can this method proceed even if Exceptions are thrown in the upload?  Or should those be thrown up to the user?

Reply:
Here we are giving a chance to all panels (steps) to interact with the process. If a custom panel fails for some external condition the upload can still proceed... I'm not sure that is the best way to proceed here but I see similar behaviour in dspace consumers often used as extension point.